apiserver/facades/controller/caasunitprovisioner/provisioner.go

```diff
@@ -0,0 +1,206 @@
+	results := params.StringsWatchResults{
+		Results: make([]params.StringsWatchResult, len(args.Entities)),
+	}
+	for i, arg := range args.Entities {
+		id, changes, err := f.watchUnits(arg.Tag)
+		if err != nil {
+			results.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		results.Results[i].StringsWatcherId = id
+		results.Results[i].Changes = changes
+	}
+	return results, nil
+}
+
+func (f *Facade) watchUnits(tagString string) (string, []string, error) {
+	tag, err := names.ParseApplicationTag(tagString)
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+	app, err := f.state.Application(tag.Id())
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+	w := app.WatchUnits()
+	if changes, ok := <-w.Changes(); ok {
+		return f.resources.Register(w), changes, nil
+	}
+	return "", nil, watcher.EnsureErr(w)
+}
+
+// WatchContainerSpec starts a NotifyWatcher to watch changes to the
+// container spec for specified units in this model.
+func (f *Facade) WatchContainerSpec(args params.Entities) (params.NotifyWatchResults, error) {
+	model, err := f.state.Model()
+	if err != nil {
+		return params.NotifyWatchResults{}, errors.Trace(err)
+	}
+	results := params.NotifyWatchResults{
+		Results: make([]params.NotifyWatchResult, len(args.Entities)),
+	}
+	for i, arg := range args.Entities {
+		id, err := f.watchContainerSpec(model, arg.Tag)
+		if err != nil {
+			results.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		results.Results[i].NotifyWatcherId = id
+	}
+	return results, nil
+}
+
+func (f *Facade) watchContainerSpec(model Model, tagString string) (string, error) {
+	tag, err := names.ParseUnitTag(tagString)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	w, err := model.WatchContainerSpec(tag)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	if _, ok := <-w.Changes(); ok {
+		return f.resources.Register(w), nil
+	}
+	return "", watcher.EnsureErr(w)
+}
+
+// ContainerSpec returns the container spec for specified units in this model.
+func (f *Facade) ContainerSpec(args params.Entities) (params.StringResults, error) {
+	model, err := f.state.Model()
+	if err != nil {
+		return params.StringResults{}, errors.Trace(err)
+	}
+	results := params.StringResults{
+		Results: make([]params.StringResult, len(args.Entities)),
+	}
+	for i, arg := range args.Entities {
+		spec, err := f.containerSpec(model, arg.Tag)
+		if err != nil {
+			results.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		results.Results[i].Result = spec
+	}
+	return results, nil
+}
+
+func (f *Facade) containerSpec(model Model, tagString string) (string, error) {
+	tag, err := names.ParseUnitTag(tagString)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return model.ContainerSpec(tag)
+}
+
+// Life returns the lifecycle state for specified applications and units
+// in this model.
+func (f *Facade) Life(args params.Entities) (params.LifeResults, error) {
```

Comment:
can we use the common ewLifeGetter plugin here - we'll need to tweak the method below to be an EntityFinder

Reply:
done